pkg/agent/route/route_linux.go

```diff
@@ -388,6 +389,96 @@ func (c *Client) writeEKSMangleRule(iptablesData *bytes.Buffer) {
 	}...)
 }
 
+// ensureEKSConnmarkRule ensures that an extra iptables rule is present in the
+// nat table. This rule is required to mark Pod traffic, so that it is routed
+// using the correct route table. It is especially important for Pods getting an
+// IP address from a secondary network interface (ENI), as they are routed with
+// a different route table by default (not "main"). Without this rule, these
+// Pods don't have connectivity to the Internet.
+// Note that this rule is only needed for IPv4.
+// See https://docs.aws.amazon.com/eks/latest/userguide/external-snat.html for
+// more details.
+func (c *Client) ensureEKSConnmarkRule() error {
+	klog.V(2).InfoS("Ensure extra iptables nat rule for EKS is present so that Pod traffic is marked correctly")
+	existingRules, err := c.ipt.ListRules(iptables.NATTable, iptables.PreRoutingChain)
+	if err != nil {
+		return err
+	}
+
+	ruleSpecToRule := func(ruleSpec []string) string {
+		rule := make([]string, 0, len(ruleSpec))
+		for _, s := range ruleSpec {
+			if strings.Contains(s, " ") {
+				rule = append(rule, fmt.Sprintf(`"%s"`, s))
+			} else {
+				rule = append(rule, s)
+			}
+		}
+		return strings.Join(rule, " ")
+	}
+
+	awsRuleSpec := []string{
+		"-i", "eni+",
+		"-m", "comment", "--comment", "AWS, outbound connections",
+		"-m", "state", "--state", "NEW",
+		"-j", "AWS-CONNMARK-CHAIN-0",
+	}
+	awsRule := ruleSpecToRule(awsRuleSpec)
+	antreaRuleSpec := []string{
+		"-i", c.nodeConfig.GatewayConfig.Name,
+		"-m", "comment", "--comment", "Antrea: AWS, outbound connections",
+		"-m", "state", "--state", "NEW",
+		"-j", "AWS-CONNMARK-CHAIN-0",
+	}
+	antreaRule := ruleSpecToRule(antreaRuleSpec)
+	var awsRuleIdx, antreaRuleIdx int
+	// start at 1 to skip the default policy rule for the chain
```

Comment:
It seems a little complicated to insert/delete the rule, could we append an antrea prerouting chain like other antrea rules and use iptables-restore to insert or delete the rule according to presense of the target aws rule?

I think we don't really need to insert the rule before `-A PREROUTING -m comment --comment "AWS, CONNMARK" -j CONNMARK --restore-mark --nfmask 0x80 --ctmask 0x80` because only the first packet in a connection will be evaluated in nat table, so in order to mark the subsequent packets, there must be same rule installed in another table (perhaps mangle), and this rule in nat table seems redundant and works for the first packet only. The condition "-m state --state NEW" seems useless too but I understand it's to keep consistency with aws rule.

Reply:
I agree with you. And they do have the correct rule in the mangle table:
```
-A PREROUTING -i eni+ -m comment --comment "AWS, primary ENI" -j CONNMARK --restore-mark --nfmask 0x80 --ctmask 0x80
```
The problem is that this rule is only added when `AWS_VPC_CNI_NODE_PORT_SUPPORT` is set to true (the default), while it should be added when `AWS_VPC_CNI_NODE_PORT_SUPPORT` is set to true or `AWS_VPC_K8S_CNI_EXTERNALSNAT` is set to false (also the default).

See https://github.com/aws/amazon-vpc-cni-k8s/blob/f979630f9d740bbc0ef26135b05ccffdd4ab841f/pkg/networkutils/network.go#L529-L538

I think I'll try to improve the approach here, as you suggested, and I'll also see if I can submit a patch for amazon-vpc-cni to:
1) remove the `--restore-mark` rule in the nat table
2) fix the condition for the mangle rule
3) remove the matching on `--state NEW` in the nat table

Reply:
I followed your suggestion. The `-A PREROUTING -m comment --comment "AWS, CONNMARK" -j CONNMARK --restore-mark --nfmask 0x80 --ctmask 0x80` actually serves a purpose. While there is a similar rule in the mangle table, it will not apply to the first packet in the connection, given that the mangle table is traversed before the nat table for prerouting.

Reply:
The rule in mangle table has the condition `-i eni+`, I'm curious how it applies to traffic coming to the host network namespace via `antrea-gw0`, or it's because of something else?

Reply:
The default mangle rule for AWS is `-A PREROUTING -i eni+ -m comment --comment "AWS, primary ENI" -j CONNMARK --restore-mark --nfmask 0x80 --ctmask 0x80`, which I think is the rule that you are referring to. But we also add the following Antrea rule for AWS:
`-A ANTREA-MANGLE -i antrea-gw0 -m comment --comment "Antrea: AWS, primary ENI" -j CONNMARK --restore-mark --nfmask 0x80 --ctmask 0x80`.

I added that rule in the past for a separate, but related issue, which affected NodePort support. The rule also applies in this case, so I don't need to add a new one. Hopefully my comments for `writeEKSMangleRules` explain this.

Reply:
Thanks for the explanation! I was struggling to understand how handling only the first packet fixed it. I should have taken a look at Antrea's rules.